mutiny-core/src/blindauth.rs

```diff
@@ -0,0 +1,332 @@
+    }
+
+    pub async fn used_token(&self, token: BlindedToken) -> Result<(), MutinyError> {
+        // once a token has sufficiently been used, mark it as spent and save it back
+        let mut token_storage_guard = self.token_storage.write().await;
+
+        // find the token in the vector of tokens
+        if let Some(index) = token_storage_guard.tokens.iter_mut().position(|t| {
+            t.service_id == token.service_id
+                && t.plan_id == token.plan_id
+                && t.counter == token.counter
+        }) {
+            // mark the found token as spent
+            token_storage_guard.tokens[index].spent = true;
+            token_storage_guard.version += 1;
+
+            // save the updated token storage back to the database or other persistent storage
+            self.storage
+                .insert_token_storage(token_storage_guard.clone())
+                .await?;
+        } else {
+            return Err(MutinyError::NotFound);
+        }
+
+        Ok(())
+    }
+}
+
+async fn get_available_tokens(
+    auth_client: &MutinyAuthClient,
+    base_url: &str,
+) -> Result<CheckServiceTokenResponse, MutinyError> {
+    let url = Url::parse(&format!("{}/v1/check-tokens", base_url))
+        .map_err(|_| MutinyError::ConnectionFailed)?;
+    let res = auth_client
+        .request(Method::POST, url, None)
+        .await?
+        .json::<CheckServiceTokenResponse>()
+        .await
+        .map_err(|_| MutinyError::ConnectionFailed)?;
+
+    Ok(res)
+}
+
+async fn post_redeem_tokens(
+    auth_client: &MutinyAuthClient,
+    base_url: &str,
+    req: RedeemServiceTokenRequest,
+) -> Result<RedeemServiceTokenResponse, MutinyError> {
+    let url = Url::parse(&format!("{}/v1/redeem-tokens", base_url))
+        .map_err(|_| MutinyError::ConnectionFailed)?;
+    let body = serde_json::to_value(req)?;
+    let res = auth_client
+        .request(Method::POST, url, Some(body))
+        .await?
+        .json::<RedeemServiceTokenResponse>()
+        .await
+        .map_err(|_| MutinyError::ConnectionFailed)?;
+
+    Ok(res)
+}
+
+async fn derive_blind_token(
+    secret: &DerivableSecret,
+    service_id: u32,
+    plan_id: u32,
+    counter: u32,
+) -> Result<NotYetBlindedToken, MutinyError> {
+    let child_secret = secret
+        .child_key(SERVICE_REGISTRATION_CHILD_ID)
+        .child_key(ChildId(service_id.into()))
+        .child_key(ChildId(plan_id.into()))
+        .child_key(ChildId(counter.into()));
+
+    let spend_key = child_secret
+        .child_key(SPEND_KEY_CHILD_ID)
+        .to_secp_key(fedimint_ln_common::bitcoin::secp256k1::SECP256K1);
+
+    let nonce = fedimint_mint_client::Nonce(spend_key.public_key());
+    let blinding_key = BlindingKey(
+        child_secret
+            .child_key(BLINDING_KEY_CHILD_ID)
+            .to_bls12_381_key(),
+    );
+    let blinded_message = blind_message(nonce.to_message(), blinding_key);
+
+    let blinded_token = NotYetBlindedToken {
```

Comment:
`NotYetBlindedToken` doesn't really make sense here, in the line above you are doing the blinding. Something like `UnsignedToken`  would make more sense